src/Symfony/Component/Finder/Iterator/VcsIgnoredFilterIterator.php

```diff
@@ -36,41 +41,115 @@ public function accept(): bool
     {
         $file = $this->current();
 
-        $fileRealPath = $file->getRealPath();
-        if ($file->isDir()) {
+        $fileRealPath = $this->normalizePath($file->getRealPath());
+
+        return !$this->isIgnored($fileRealPath);
+    }
+
+    private function isIgnored(string $fileRealPath): bool
+    {
+        if (is_dir($fileRealPath) && !str_ends_with($fileRealPath, '/')) {
             $fileRealPath .= '/';
         }
 
+        if (isset($this->ignoredPathsCache[$fileRealPath])) {
+            return $this->ignoredPathsCache[$fileRealPath];
+        }
+
+        $ignored = false;
+
+        foreach ($this->parentsDirectoryDownward($fileRealPath) as $parentDirectory) {
+            if ($this->isIgnored($parentDirectory)) {
+                $ignored = true;
+
+                // rules in ignored directories are ignored, no need to check further.
+                break;
+            }
+
+            $fileRelativePath = substr($fileRealPath, \strlen($parentDirectory) + 1);
+
+            $regexps = $this->readGitignoreFile("{$parentDirectory}/.gitignore");
+
+            if (null === $regexps) {
+                continue;
+            }
+
+            [$exclusionRegex, $inclusionRegex] = $regexps;
+
+            if (preg_match($exclusionRegex, $fileRelativePath)) {
+                $ignored = true;
+
+                continue;
+            }
+
+            if (preg_match($inclusionRegex, $fileRelativePath)) {
+                $ignored = false;
+            }
+        }
+
+        return $this->ignoredPathsCache[$fileRealPath] = $ignored;
+    }
+
+    /**
+     * @return list<string>
+     */
+    private function parentsDirectoryDownward(string $fileRealPath): array
+    {
+        $parentDirectories = [];
+
         $parentDirectory = $fileRealPath;
 
-        do {
-            $parentDirectory = \dirname($parentDirectory);
-            $relativeFilePath = substr($fileRealPath, \strlen($parentDirectory) + 1);
+        while (true) {
+            $newParentDirectory = \dirname($parentDirectory);
+
+            // dirname('/') = '/'
+            if ($newParentDirectory === $parentDirectory) {
+                break;
+            }
 
-            $regex = $this->readGitignoreFile("{$parentDirectory}/.gitignore");
+            $parentDirectory = $newParentDirectory;
 
-            if (null !== $regex && preg_match($regex, $relativeFilePath)) {
-                return false;
+            if (0 !== strpos($parentDirectory, $this->baseDir)) {
```

Comment:
To fully support the `git` behaviour, we should search up too the filesystem root or until the `.git` subdirectory is found. If there is no such directory, keep the current behaviour, eg. stop at the baseDir.

Reply:
I think this is something we want to avoid (see https://github.com/symfony/symfony/pull/43150#issuecomment-929263087).

Reply:
I mean to TRY to find the `.git` root directory. By TRY I mean to ignore possible/expected errors when search above the baseDir.

This is git behaviour and we should support this. Maybe with a switch to disable it, but it should be enabled by default as per git standard.